design/StableMemory.md

```diff
@@ -0,0 +1,88 @@
+# Stable Memory API
+
+The current implementation of stable variables is based on
+serialization and deserialization of all stable data on upgrade. This
+clearly doesn't scale to large amounts of stable data as there may not
+be enough cycles to perform (de)serialization.
+
+To avoid this upgrade hazard, some Rust canisters with low-level API
+access, and large stable memory footprints arrange to store their
+persistent data in stable memory at all times, using either a custom
+binary encoding or a mixture of candid and raw binary.
+
+To provide more fine-grained access to stable memory we propose:
+
+A) extending Motoko with access to (monomorphic) Candid serialization primitives.
+   This is most easily achieved by extending the syntax with saturated serialization and deserialization
+   constructs, along the lines of debug_show.
+
+```
+   encode<T>(v : T) : Blob
+   decode<T>(b : Blob) : T
+
+   (where T shared or, to allow isolated graphs, T stable).
+```
+
+The current implementations of these prims embed the type table with the value. This means blobs are larger, but also self-describing, allowing per blob upgrade.
+
+
+
+B) Surfacing stable memory API
+
+```
+ic0.stable_size : () -> (page_count : i32);                                 // *
+ic0.stable_grow : (new_pages : i32) -> (old_page_count : i32);              // *
+ic0.stable_write : (offset : i32, src : i32, size : i32) -> ();             // *
+ic0.stable_read : (dst : i32, offset : i32, size : i32) -> ();              // *
+```
+
+These grow memory and do bulk transfers between Wasm and stable memory.
+The `// *` means that they can be called in all contexts (e.g. init, update, query etc).
+Direct reads and writes of word-sized data to/from the stack are not supported but can be emulated at cost.
+
+The Minimal Motoko prims could be:
+
+```
+module StableMemory {
+  stableSize : () -> (page_count : Nat32); // hidden?
+  stableGrow : (new_pages : Nat32) -> (old_page_count : Nat32); // hidden?
+  stableWrite : (offset : Nat32, b : Blob) -> (); // write size + contents of blob to memory, growing stable memory region on demand
+  stableRead : (offset : Nat32) -> Blob // read Blob size & contents from memory at offset into fresh blob, trapping if out-of-bounds
```

Comment:
Hm, this interface involves a substantial allocation and de/serialisation overhead for every single access to stable memory. That would be pretty expensive for some needs.

But more importantly, the serialisation format produces blobs whose size are variable not just by type but also by value. That makes it practically impossible to update any data in place with just this API.

In order to provide sufficient control over memory layout, I think we need something more low-level, that e.g. enables reading/writing basic types like numbers (and perhaps vectors thereof) directly, with fixed size.

We can still provide the variable-size blob functions, but I suspect their utility is limited to cases where the app is fine with essentially custom-implementing full dynamic memory management inside stable memory (which still requires fixed-size read/writes to boot).

Reply:
In that case, maybe we should provide just raw word level memory access, and let the user write custom picklers or whatever. 
That pretty low-level though.

The internet identity service actually uses the technique of an array fixed size regions, with candid serialization in and out of each entry. That supports a modicum of lazy upgrade, but writes can fail if they exceed the slot budget.

Reply:
> but I suspect their utility is limited to cases where the app is fine with essentially custom-implementing full dynamic memory management inside stable memory (which still requires fixed-size read/writes to boot).

It seems very limited or bad when stated that way (which is a fair way to state it), but still, I think most apps will have a data model that makes that problem very domain-specific, and thus, possibly tractable to solve on a per canister basis, at least for initial cases.

For instance, the II service uses arrays of fixed-sized fields, indexed by array offset, as @crusso mentioned above. No need for general-purpose MM there.

If anything, I would imagine that apps want a lot of control over when data becomes "garbage", and may want a quasi-manual process for marking users' stable data as garbage, not something automagical or otherwise obscured by the language semantics.